core/bft_test.go

```diff
@@ -0,0 +1,790 @@
+        receiver: 0
+        seq: 0
+        result: '{"Height":1,"PrevBlockHash":null,"Payload":"dGVzdCByZXF1ZXN0IG1lc3NhZ2U="}'
+  - type: check
+    messages:
+      - type: prepare
+        cv: 1
+        replica: 0
+        view: 0
+        client: 0
+        seq: 0
+        sender: 1
+        receiver: 1
+      - type: prepare
+        cv: 1
+        replica: 0
+        view: 0
+        client: 0
+        seq: 0
+        sender: 1
+        receiver: 2
+      - type: reply
+        seq: 0
+        result: '{"Height":1,"PrevBlockHash":null,"Payload":"dGVzdCByZXF1ZXN0IG1lc3NhZ2U="}'
+        sender: 1
+        receiver: 0
+`
+
+	result := runSimulator(input)
+
+	assert.True(t, result)
+	assert.Equal(t, uint64(1), replicaStacksBft[uint32(0)].BlockHeight())
+}
+
+func testTwoClientScenarioBackup(t *testing.T) {
+	input := `
+steps:
+  - type: send
+    messages:
+      - type: request
+        sender: 0
+        operation: request from client 0
+      - type: request
+        sender: 1
+        operation: request from client 1
+      - type: prepare
+        client: 0
+        sender: 0
+        receiver: 1
+        view: 0
+        seq: 0
+        prepareid: 0
+      - type: commit
+        sender: 2
+        receiver: 1
+        prepareid: 0
+      - type: prepare
+        client: 1
+        sender: 0
+        receiver: 1
+        view: 0
+        seq: 0
+        prepareid: 1
+      - type: commit
+        sender: 2
+        receiver: 1
+        prepareid: 1
+      - type: reply
+        sender: 0
+        receiver: 0
+        seq: 0
+        result: '{"Height":1,"PrevBlockHash":null,"Payload":"dGVzdCByZXF1ZXN0IG1lc3NhZ2U="}'
+      - type: reply
+        sender: 0
+        receiver: 1
+        seq: 0
+        result: '{"Height":2,"PrevBlockHash":"6qdy4iR8cHF0jarsHSgtLuu4towvKUNrWllDw2yLarI=","Payload":"cmVxdWVzdCBmcm9tIGNsaWVudCAx"}'
+  - type: check
+    messages:
+      - type: commit
+        sender: 1
+        receiver: 2
+        cv: 1
+      - type: commit
+        sender: 1
+        receiver: 0
+        cv: 1
+      - type: commit
+        sender: 1
+        receiver: 2
+        cv: 2
+      - type: commit
+        sender: 1
+        receiver: 0
+        cv: 2
+      - type: reply
+        sender: 0
+        receiver: 0
+        seq: 0
+        result: '{"Height":1,"PrevBlockHash":null,"Payload":"dGVzdCByZXF1ZXN0IG1lc3NhZ2U="}'
+      - type: reply
+        sender: 0
+        receiver: 1
+        seq: 0
+        result: '{"Height":2,"PrevBlockHash":"6qdy4iR8cHF0jarsHSgtLuu4towvKUNrWllDw2yLarI=","Payload":"cmVxdWVzdCBmcm9tIGNsaWVudCAx"}'
+`
+
+	result := runSimulator(input)
+
+	assert.True(t, result)
+	assert.Equal(t, uint64(2), replicaStacksBft[uint32(1)].BlockHeight())
+}
+
+func TestSimulator(t *testing.T) {
+	testCases := []struct {
+		numReplica    int
+		numClient     int
+		targetReplica int
+		testcase      func(t *testing.T)
+	}{
+		{numReplica: 3, numClient: 1, targetReplica: 1, testcase: testSimpleScenarioBackup},
```

Comment:
How about moving the parameters, `numReplica`, etc. to a YAML as well as a scenario? I am thinking about aggregating a kind of test data and environment definitions to the external structure to separate the code into a test runner and data.

Reply:
Sounds nice, thanks. I'll move the parameters into YAML.